internal/pkg/policy/policy_output.go

```diff
@@ -0,0 +1,179 @@
+	switch p.Type {
+	case OutputTypeElasticsearch:
+		zlog.Debug().Msg("preparing elasticsearch output")
+
+		// The role is required to do api key management
+		if p.Role == nil {
+			zlog.Error().Str("name", p.Name).Msg("policy does not contain required output permission section")
+			return ErrNoOutputPerms
+		}
+
+		// Determine whether we need to generate an output ApiKey.
+		// This is accomplished by comparing the sha2 hash stored in the agent
+		// record with the precalculated sha2 hash of the role.
+
+		// Note: This will need to be updated when doing multi-cluster elasticsearch support
+		// There will need to be multiple api keys tracked on the agent model
+		needKey := true
+		switch {
+		case agent.DefaultApiKey == "":
+			zlog.Debug().Msg("must generate api key as default API key is not present")
+		case p.Role.Sha2 != agent.PolicyOutputPermissionsHash:
+			zlog.Debug().Msg("must generate api key as policy output permissions changed")
+		default:
+			needKey = false
+			zlog.Debug().Msg("policy output permissions are the same")
+		}
+
+		if needKey {
+			zlog.Debug().
+				RawJSON("roles", p.Role.Raw).
+				Str("oldHash", agent.PolicyOutputPermissionsHash).
+				Str("newHash", p.Role.Sha2).
+				Msg("Generating a new API key")
+
+			outputAPIKey, err := generateOutputAPIKey(ctx, bulker, agent.Id, p.Name, p.Role.Raw)
+			if err != nil {
+				zlog.Error().Err(err).Msg("fail generate output key")
+				return err
+			}
+
+			if ok := setMapObj(outputMap, outputAPIKey, p.Name, "api_key"); !ok {
+				return ErrFailInjectAPIKey
+			}
+
+			if isDefault {
+				zlog.Info().
+					Str("hash.sha256", p.Role.Sha2).
+					Str(logger.DefaultOutputApiKeyId, outputAPIKey.Id).
+					Msg("Updating agent record to pick up default output key.")
+
+				fields := map[string]interface{}{
+					dl.FieldDefaultApiKey:               outputAPIKey.Agent(),
+					dl.FieldDefaultApiKeyId:             outputAPIKey.Id,
+					dl.FieldPolicyOutputPermissionsHash: p.Role.Sha2,
+				}
+				if agent.DefaultApiKeyId != "" {
+					fields[dl.FieldDefaultApiKeyHistory] = model.DefaultApiKeyHistoryItems{
+						Id:        agent.DefaultApiKeyId,
+						RetiredAt: time.Now().UTC().Format(time.RFC3339),
+					}
+				}
+				body, err := renderUpdatePainlessScript(fields)
+
+				// Using painless script to append the old keys to the history
+
+				if err != nil {
+					return err
+				}
+
+				if err = bulker.Update(ctx, dl.FleetAgents, agent.Id, body); err != nil {
+					zlog.Error().Err(err).Msg("fail update agent record")
+					return err
+				}
+				agent.DefaultApiKey = outputAPIKey.Agent()
+			}
+		}
+	case OutputTypeLogstash:
+		zlog.Debug().Msg("preparing logstash output")
+		zlog.Info().Msg("no actions required for logstash output preparation")
+	default:
+		zlog.Debug().Msgf("unknown output type: %s; skipping preparation", p.Type)
```

Comment:
Not sure what was the previous behavior, but this looks like we should _return an error_ and log it as an error. I supposed the only way to get into that section of the code is that the format received from Fleet is incorrect.